src/doc/trpl/error-handling.md

```diff
@@ -1,308 +1,2130 @@
-        return Err(e)
+Notice that we still have the calls to `map_err`. Why? Well, recall the
+definitions of [`try!`](#code-try-def) and [`From`](#code-from-def). The
+problem is that there is no `From` impl that allows us to convert from error
+types like `io::Error` and `num::ParseIntError` to our own custom `CliError`.
+Of course, it is easy to fix this! Since we defined `CliError`, we can impl
+`From` with it:
+
+```rust
+# #[derive(Debug)]
+# enum CliError { Io(io::Error), Parse(num::ParseIntError) }
+use std::io;
+use std::num;
+
+impl From<io::Error> for CliError {
+    fn from(err: io::Error) -> CliError {
+        CliError::Io(err)
     }
+}
 
-    return Ok(());
+impl From<num::ParseIntError> for CliError {
+    fn from(err: num::ParseIntError) -> CliError {
+        CliError::Parse(err)
+    }
 }
 ```
 
-With this:
+All these impls are doing is teaching `From` how to create a `CliError` from
+other error types. In our case, construction is as simple as invoking the
+corresponding value constructor. Indeed, it is *typically* this easy.
+
+We can finally rewrite `file_double`:
 
 ```rust
+# use std::io;
+# use std::num;
+# enum CliError { Io(::std::io::Error), Parse(::std::num::ParseIntError) }
+# impl From<io::Error> for CliError {
+#     fn from(err: io::Error) -> CliError { CliError::Io(err) }
+# }
+# impl From<num::ParseIntError> for CliError {
+#     fn from(err: num::ParseIntError) -> CliError { CliError::Parse(err) }
+# }
+
 use std::fs::File;
+use std::io::Read;
+use std::path::Path;
+
+fn file_double<P: AsRef<Path>>(file_path: P) -> Result<i32, CliError> {
+    let mut file = try!(File::open(file_path));
+    let mut contents = String::new();
+    try!(file.read_to_string(&mut contents));
+    let n: i32 = try!(contents.trim().parse());
+    Ok(2 * n)
+}
+```
+
+The only thing we did here was remove the calls to `map_err`. They are no
+longer needed because the `try!` macro invokes `From::from` on the error value.
+This works because we've provided `From` impls for all the error types that
+could appear.
+
+If we modified our `file_double` function to perform some other operation, say,
+convert a string to a float, then we'd need to add a new variant to our error
+type:
+
+```rust
 use std::io;
-use std::io::prelude::*;
+use std::num;
+
+enum CliError {
+    Io(io::Error),
+    ParseInt(num::ParseIntError),
+    ParseFloat(num::ParseFloatError),
+}
+```
 
-struct Info {
-    name: String,
-    age: i32,
-    rating: i32,
+And add a new `From` impl:
+
+```rust
+# enum CliError {
+#     Io(::std::io::Error),
+#     ParseInt(num::ParseIntError),
+#     ParseFloat(num::ParseFloatError),
+# }
+
+use std::num;
+
+impl From<num::ParseFloatError> for CliError {
+    fn from(err: num::ParseFloatError) -> CliError {
+        CliError::ParseFloat(err)
+    }
 }
+```
+
+And that's it!
+### Advice for library writers
+
+Idioms for Rust libraries are still forming, but if your library needs
```

Comment:
Text like this gets out of date quickly, can we just start with "If your library needs to report"